File: Python/maximum-frequency-score-of-a-subarray.py
# Time:  O(n)
# Space: O(n)

import collections


# two pointers, freq table, hash table
class Solution(object):
    def maxFrequencyScore(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        MOD = 10**9+7
        def powmod(n, p):
            if (n, p) not in lookup:
                lookup[n, p] = (lookup[n, p-1]*n)%MOD if p >= 2 else n%MOD
            return lookup[n, p]

        cnt = collections.Counter()
        lookup = {}
        result = curr = 0
        for i in xrange(len(nums)):
            if i >= k:
                curr = (curr-powmod(nums[i-k], cnt[nums[i-k]]))%MOD
                cnt[nums[i-k]] -= 1
                if cnt[nums[i-k]]:
                    curr = (curr+powmod(nums[i-k], cnt[nums[i-k]]))%MOD
            if cnt[nums[i]]:
               curr = (curr-powmod(nums[i], cnt[nums[i]]))%MOD
            cnt[nums[i]] += 1
            curr = (curr+powmod(nums[i], cnt[nums[i]]))%MOD
            if i >= k-1:
                result = max(result, curr)
        return result


# Time:  O(nlogn)
# Space: O(n)
import collections


# two pointers, freq table
class Solution2(object):
    def maxFrequencyScore(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        MOD = 10**9+7
        cnt = collections.Counter()
        result = curr = 0
        for i in xrange(len(nums)):
            if i >= k:
                curr = (curr-pow(nums[i-k], cnt[nums[i-k]], MOD))%MOD
                cnt[nums[i-k]] -= 1
                if cnt[nums[i-k]]:
                    curr = (curr+pow(nums[i-k], cnt[nums[i-k]], MOD))%MOD
            if cnt[nums[i]]:
               curr = (curr-pow(nums[i], cnt[nums[i]], MOD))%MOD
            cnt[nums[i]] += 1
            curr = (curr+pow(nums[i], cnt[nums[i]], MOD))%MOD
            if i >= k-1:
                result = max(result, curr)
        return result

File: C++/maximum-frequency-score-of-a-subarray.cpp
// Time:  O(n)
// Space: O(n)

// two pointers, sliding window, freq table, hash table
class Solution {
public:
    int maxFrequencyScore(vector<int>& nums, int k) {
        static const int MOD = 1e9 + 7;
        unordered_map<int, unordered_map<int, int>> lookup;
        const auto& powmod = [&](int n, int p) {
            if (!lookup.count(n) || !lookup[n].count(p)) {
                lookup[n][p] = p >= 2 ? (static_cast<int64_t>(lookup[n][p - 1]) * n) % MOD : n % MOD;  // assumed powmod(n, p-1) was accessed before powmod(n, p)
            }
            return lookup[n][p];
        };

        int result = 0, curr = 0;
        unordered_map<int, int> cnt;
        for (int i = 0; i < size(nums); ++i) {
            if (i >= k) {
                curr = ((curr - powmod(nums[i - k], cnt[nums[i - k]])) % MOD + MOD) % MOD;
                --cnt[nums[i - k]];
                if (cnt[nums[i - k]]) {
                    curr = (curr + powmod(nums[i - k], cnt[nums[i - k]])) % MOD;
                }
            }
            if (cnt[nums[i]]) {
                curr = ((curr - powmod(nums[i], cnt[nums[i]])) % MOD + MOD) % MOD;
            }
            ++cnt[nums[i]];
            curr = (curr + powmod(nums[i], cnt[nums[i]])) % MOD;
            if (i >= k - 1) {
                result = max(result, curr);
            }
        }
        return result;
    }
};


// Time:  O(n)
// Space: O(n)
// two pointers, sliding window, freq table
class Solution2 {
public:
    int maxFrequencyScore(vector<int>& nums, int k) {
        static const int MOD = 1e9 + 7;
        const auto& powmod = [&](int a, int b) {
            a %= MOD;
            int64_t result = 1;
            while (b) {
                if (b & 1) {
                    result = result * a % MOD;
                }
                a = int64_t(a) * a % MOD;
                b >>= 1;
            }
            return result;
        };

        int result = 0, curr = 0;
        unordered_map<int, int> cnt;
        for (int i = 0; i < size(nums); ++i) {
            if (i >= k) {
                curr = ((curr - powmod(nums[i - k], cnt[nums[i - k]])) % MOD + MOD) % MOD;
                --cnt[nums[i - k]];
                if (cnt[nums[i - k]]) {
                    curr = (curr + powmod(nums[i - k], cnt[nums[i - k]])) % MOD;
                }
            }
            if (cnt[nums[i]]) {
                curr = ((curr - powmod(nums[i], cnt[nums[i]])) % MOD + MOD) % MOD;
            }
            ++cnt[nums[i]];
            curr = (curr + powmod(nums[i], cnt[nums[i]])) % MOD;
            if (i >= k - 1) {
                result = max(result, curr);
            }
        }
        return result;
    }
};
